libyul/optimiser/RedundantAssignEliminator.cpp

```diff
@@ -54,48 +54,50 @@ void RedundantAssignEliminator::operator()(Assignment const& _assignment)
 
 	if (_assignment.variableNames.size() == 1)
 		// Default-construct it in "Undecided" state if it does not yet exist.
-		m_assignments[_assignment.variableNames.front().name][&_assignment];
+		m_tracker.assignments[_assignment.variableNames.front().name][&_assignment];
 }
 
 void RedundantAssignEliminator::operator()(If const& _if)
 {
 	visit(*_if.condition);
 
-	RedundantAssignEliminator branch{*this};
-	branch(_if.body);
+	TrackingState branch{m_tracker};
+	(*this)(_if.body);
+	swap(m_tracker, branch);
 
-	join(branch);
+	join(move(branch));
 }
 
 void RedundantAssignEliminator::operator()(Switch const& _switch)
 {
 	visit(*_switch.expression);
 
 	bool hasDefault = false;
-	vector<RedundantAssignEliminator> branches;
+	vector<TrackingState> branches;
 	for (auto const& c: _switch.cases)
 	{
 		if (!c.value)
 			hasDefault = true;
-		branches.emplace_back(*this);
-		branches.back()(c.body);
+		branches.emplace_back(m_tracker);
```

Comment:
Can the swap be avoided by storing the pre-switch-state in a local variable before entering the loop over the cases?

Reply:
Yup.